plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/object/struct/EditForeignKeyPage.java

```diff
@@ -1013,55 +1014,83 @@ private void setReferenceAttributes(Collection<? extends DBSEntityAttribute> ref
 
     @Nullable
     public static DBVEntityForeignKey createVirtualForeignKey(@NotNull DBVEntity vEntity) {
-        return createVirtualForeignKey(vEntity, null, new FKType[] {FK_TYPE_LOGICAL}, null, null);
+        return makeVirtualForeignKeySupplier(vEntity, null, new FKType[] {FK_TYPE_LOGICAL}, null, null).get();
     }
-
+    
     @Nullable
-    public static DBVEntityForeignKey createVirtualForeignKey(
+    public static Supplier<DBVEntityForeignKey> makeVirtualForeignKeySupplier(
         @NotNull DBVEntity vEntity,
         @Nullable DBSEntity refEntity,
         @Nullable FKType[] allowedKeyTypes,
         @Nullable Collection<? extends DBSEntityAttribute> srcAttributes,
         @Nullable Collection<? extends DBSEntityAttribute> refAttributes)
     {
-        DBVEntityForeignKey virtualFK = new DBVEntityForeignKey(vEntity);
-        EditForeignKeyPage editDialog = new EditForeignKeyPage(
-            "Define virtual foreign keys",
-            virtualFK,
-            new DBSForeignKeyModifyRule[]{DBSForeignKeyModifyRule.NO_ACTION},
-            Collections.emptyMap());
-        editDialog.setEnableCustomKeys(true);
-        if (allowedKeyTypes != null) {
-            editDialog.setAllowedKeyTypes(allowedKeyTypes);
-        }
-        if (refEntity != null) {
-            editDialog.setRefTable(refEntity);
-        }
-        if (srcAttributes != null) {
-            editDialog.setSourceAttributes(srcAttributes);
-        }
-        if (refAttributes != null) {
-            editDialog.setReferenceAttributes(refAttributes);
-        }
-        if (!editDialog.edit()) {
-            return null;
-        }
-        // Save
-        try {
-            virtualFK.setReferencedConstraint(new VoidProgressMonitor(), editDialog.getUniqueConstraint());
-        } catch (DBException e1) {
-            log.error(e1);
-            return null;
-        }
-        List<DBVEntityForeignKeyColumn> columns = new ArrayList<>();
-        for (FKColumnInfo tableColumn : editDialog.getColumns()) {
-            columns.add(
-                new DBVEntityForeignKeyColumn(
-                    virtualFK, tableColumn.getOwnColumn().getName(), tableColumn.getRefColumn().getName()));
-        }
-        virtualFK.setAttributes(columns);
-        vEntity.addForeignKey(virtualFK);
-        return virtualFK;
+        return new Supplier<DBVEntityForeignKey>() {
+            private List<FKColumnInfo> tableColumns = null;
+            private DBSEntityConstraint constraint = null;
```

Comment:
Why are they fields? They don't seem to be used anywhere but in the `get` method.

Reply:
These fields store information. They are read and changed between calls.